File: sortMatrixByDiagonals.py
from typing import List
from collections import defaultdict

class Solution:
    def sortMatrix(self, grid: List[List[int]]) -> List[List[int]]:
        n = len(grid)
        diags = defaultdict(list)

        # Collect values by diagonal (i - j)
        for i in range(n):
            for j in range(n):
                diags[i - j].append(grid[i][j])

        # Sort diagonals depending on region
        for d in diags:
            if d >= 0:  # bottom-left incl main diag
                diags[d].sort(reverse=True)  # non-increasing
            else:       # top-right
                diags[d].sort()              # non-decreasing

        # Write back to grid
        for i in range(n):
            for j in range(n):
                grid[i][j] = diags[i - j].pop(0)

        return grid

File: sortMatrixByDiagonals.go
package main

import (
	"fmt"
	"sort"
)

func sortMatrix(grid [][]int) [][]int {
	n := len(grid)
	diags := map[int][]int{}

	// Collect values by diagonal (i - j)
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			d := i - j
			diags[d] = append(diags[d], grid[i][j])
		}
	}

	// Sort diagonals depending on region
	for d, vals := range diags {
		if d >= 0 {
			sort.Sort(sort.Reverse(sort.IntSlice(vals))) // non-increasing
		} else {
			sort.Ints(vals) // non-decreasing
		}
		diags[d] = vals
	}

	// Write back to grid
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			d := i - j
			grid[i][j] = diags[d][0]
			diags[d] = diags[d][1:]
		}
	}

	return grid
}

/*
// testing
func main() {
	grid := [][]int{{1, 7, 3}, {9, 8, 2}, {4, 5, 6}}
	fmt.Println(sortMatrix(grid))
}
*/
